### ISSUE TITLE
Question about mimi codec in generator.py

### ISSUE THREAD

**Issue by tibrisch** [SesameAILabs/csm#105]
I am trying to figure out how to use the decoder so that I can stream 80ms frames, instead of decoding everything at the end.
I made a small change in generator.py

        for _ in range(max_audio_frames):
            sample = self._model.generate_frame(curr_tokens, curr_tokens_mask, curr_pos, temperature, topk)
            if torch.all(sample == 0):
                break  # eos
            yield self._audio_tokenizer.decode(torch.stack([sample]).permute(1, 2, 0)).squeeze(0).squeeze(0)

            curr_tokens = torch.cat([sample, torch.zeros(1, 1).long().to(self.device)], dim=1).unsqueeze(1)
            curr_tokens_mask = torch.cat(
                [torch.ones_like(sample).bool(), torch.zeros(1, 1).bool().to(self.device)], dim=1
            ).unsqueeze(1)
            curr_pos = curr_pos[:, -1:] + 1

However when i concatenate them later, the audio frames do not align and there are clics in the audio.
Am I doing something wrong? Or do i need to post-process the audio frames so that they align?

As a workaround I am buffering 8 frames (1920*8) 8*80 = 640ms and running _audio_tokenizer.decode on those before passing them on. That makes the clicks less annoying, but it would be nice to have 80ms audio frames.

**Comment by tibrisch** [comment]
I found the answer!
In a MIMI decoder example (of course)

https://github.com/kyutai-labs/moshi/blob/724f8909d7ab673d4db69822ddfe7030b858498e/scripts/mimi_streaming_test.py#L88

so i only needed to add one line to my code and now streaming MIMI decoding is flawless with CSM!

```
with self._audio_tokenizer.streaming(1):
    for _ in range(max_audio_frames):
```

Now the frames align, and the audio queue is filling up nicely. No  more clicks, only almost instant playback. 
I am running on RTX3060 and there is no problem with speed.

**Comment by DevChrisRoth** [comment]
@tibrisch : How fast is your new Code? In which time will the audio be generated now?

**Comment by tibrisch** [comment]
> [@tibrisch](https://github.com/tibrisch) : How fast is your new Code? In which time will the audio be generated now?

I have a fork with a restructured generator.py that supports streaming, that you can try out if you want. 
https://github.com/tibrisch/csm

A test for it is in:
https://github.com/tibrisch/csm/blob/main/stream_test.py

The short answer is that the generator is doing some setup the first time it is called....
So first TTS generation took ~1.8 seconds until first frame was in the queue.
Subsequent calls take  ~200 ms until first frame in queue.

I will see if we can break out the stuff that takes time the first call into an init function.

Also, I am getting stuttering if i dont prebuffer around 4 frames (4*80ms), but if you are on faster HW then maybe prebuffering can be avoided?

**Comment by tibrisch** [comment]
> Out of interest, are you running the chunks through something like ffmpeg to build an encoded stream, or is this local and you're just playing the PCM audio?

I am only playing pcm locally at this time. I am only trying trying to learn/understand.
Btw i think 100-150ms until first frame start to play is more correct for me. 

I suppose the ideal way to stream the audio, is to keep it as a MIMI stream  (and not decode to PCM on the server)... instead  decode at the client side.

**Comment by ZackHodari** [comment]
> first TTS generation took ~1.8 seconds 

You can try to add a warmup to the model loading, i.e. in `load_csm_1b` generate one sentence
